### PR TITLE
Unit test

### PR DESCRIPTION
**Pull request by yauhen07**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,3 @@
+*.iml
```
Comment (tumashchick) [inline .gitignore]: ok

File: src/test/java/com/epam/cdp/DivTest.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    @Test(dataProvider = "valuesForDiv")
+    public void firstDivSecond(long first, long second, long expected) {
+        long result = calc.div(first, second);
+        Assert.assertEquals(result, expected, "incorrect result division");
+    }
+
+    @DataProvider(name = "valuesForDiv")
+    public Object[][] dataForSum() {
+        return new Object[][]{
+                {4, 2, 2},
```
Comment (tumashchick) [inline src/test/java/com/epam/cdp/DivTest.java]: Add boundary values in all tests
Long.MAX_VALUE =  9223372036854775807
Long.MIN_VALUE = -9223372036854775808

Comment (tumashchick) [reply src/test/java/com/epam/cdp/DivTest.java]: Check how method works when value is out of boundary value

File: src/test/java/com/epam/cdp/SumTest.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    @Test(dataProvider = "valuesForSumTest")
+    public void FirstPlusSecond(long first, long second, long expected) {
+        long result = calc.sum(first, second);
+        Assert.assertEquals(result, expected, "invalid result sum");
+    }
+
+    @DataProvider(name = "valuesForSumTest")
+    public Object[][] dataForSum (){
+        return new Object[][]{
+                {1,2,3},
```
Comment (tumashchick) [inline src/test/java/com/epam/cdp/SumTest.java]: apply code format

File: pom.xml [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+  <artifactId>calc</artifactId>
+  <version>1.0-SNAPSHOT</version>
+
+  <name>calc</name>
+  <!-- FIXME change it to the project's website -->
+  <url>http://www.example.com</url>
+
+  <properties>
+    <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+    <maven.compiler.source>1.7</maven.compiler.source>
```
Comment (tumashchick) [inline pom.xml]: 1.10
1.7 will use java 7

File: src/test/java/com/epam/cdp/SumTest.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+import org.testng.Assert;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Parameters;
+import org.testng.annotations.Test;
+
+public class SumTest extends BasicTest {
+
+    @Test(groups = "Traning")
+    public void onePlusTwo(){
```
Comment (tumashchick) [inline src/test/java/com/epam/cdp/SumTest.java]: no need in this as u have FirstPlusSecond test

File: src/test/resources/testng.xml [hunk 6]
```diff
@@ -0,0 +1,23 @@
+<!DOCTYPE suite SYSTEM "http://testng.org/testng-1.0.dtd">
+<suite name="Demo Suite" parallel="tests" thread-count="7">
+<groups>
+    <run>
+        <exclude name="Traning"/>
+    </run>
+</groups>
+    <!--<test name="Sub Calc Test">-->
```
Comment (tumashchick) [inline src/test/resources/testng.xml]: decide leave or remove this in final version.

File: src/test/java/com/epam/cdp/MultTest.java [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+import org.testng.Assert;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Parameters;
+import org.testng.annotations.Test;
+
+public class MultTest extends BasicTest {
+
+    @Test(groups = "Traning")
+    public void zeroMultOne() {
```
[the remark this answers is not in the document]
Comment (yauhen07) [reply src/test/java/com/epam/cdp/MultTest.java]: how?

File: pom.xml [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+
+  <groupId>com.epam.cdp</groupId>
+  <artifactId>calc</artifactId>
+  <version>1.0-SNAPSHOT</version>
+
+  <name>calc</name>
+  <!-- FIXME change it to the project's website -->
+  <url>http://www.example.com</url>
```
Comment (tumashchick) [inline pom.xml]: u can remove this

File: src/test/java/com/epam/cdp/BasicTest.java [hunk 9]
```diff
@@ -0,0 +1,23 @@
+package com.epam.cdp;
+
+import com.epam.tat.module4.Calculator;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+
+public class BasicTest {
```
[the remark this answers is not in the document]
Comment (yauhen07) [reply src/test/java/com/epam/cdp/BasicTest.java]: :)

File: src/test/java/com/epam/cdp/DivTest.java [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Parameters;
+import org.testng.annotations.Test;
+
+public class DivTest extends BasicTest {
+
+    @Test(expectedExceptions = NumberFormatException.class)
+    public void divZero() {
+        long result = calc.div(1, 0);
+
```
[the remark this answers is not in the document]
Comment (yauhen07) [reply src/test/java/com/epam/cdp/DivTest.java]: ok

File: src/test/java/com/epam/cdp/MultTest.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (yauhen07) [reply src/test/java/com/epam/cdp/MultTest.java]: this test test created for traning groups, if you look at test.xml file you will see group = Training is excluded from execution

File: pom.xml (hunk 8 above)
Comment (yauhen07) [reply pom.xml]: done

File: src/test/java/com/epam/cdp/SumTest.java (hunk 5 above)
Comment (yauhen07) [reply src/test/java/com/epam/cdp/SumTest.java]: same as for firstMultSecond

File: src/test/resources/testng.xml (hunk 6 above)
Comment (yauhen07) [reply src/test/resources/testng.xml]: It is one of acceptence criteria to use Groups

File: pom.xml (hunk 4 above)
Comment (yauhen07) [reply pom.xml]: how do you know which java i'm using localy ?

File: src/test/java/com/epam/cdp/SumTest.java (hunk 3 above)
Comment (yauhen07) [reply src/test/java/com/epam/cdp/SumTest.java]: done

File: src/test/java/com/epam/cdp/junit/ExeptionsTest.java [hunk 11]
```diff
@@ -0,0 +1,11 @@
+package com.epam.cdp.junit;
+
+import org.junit.Test;
+
+public class ExeptionsTest extends BaseTestJunit {
+
+    @Test(expected = Exception.class)
```
Comment (tumashchick) [inline src/test/java/com/epam/cdp/junit/ExeptionsTest.java]: actually it is not correct to use global parent class in exception expectation as u will never know what exact type u will get

File: src/test/java/com/epam/cdp/junit/DivDoubleTest.java [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+                {0.1, 2.0, 0.05},
+                {1.0, 0.2, 5.0},
+                {0.5, 0.2, 2.5},
+                {-0.5, 0.2, -2.5},
+                {-0.5, -0.2, 2.5},
+                {0.1, 0.2, 0.5}
+        });
+    }
+
+    public DivDoubleTest(double first, double second, double expected) {
```
Comment (tumashchick) [inline src/test/java/com/epam/cdp/junit/DivDoubleTest.java]: class constructor should be after variable declaration due to code conventions. Apply to all places

File: src/test/java/com/epam/cdp/testng/PowTest.java [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+    @Parameters({"first, second, expected"})
+    @Test(dataProvider = "valuesForPow")
+    public void firstPowSecond(double first, double second, double expected) {
+        double result = calc.pow(first, second);
+        Assert.assertEquals(result, expected, "incorrect result pow");
+    }
+
+    @DataProvider(name = "valuesForPow")
+    public Object[][] dataForSum() {
+        return new Object[][]{
```
Comment (tumashchick) [inline src/test/java/com/epam/cdp/testng/PowTest.java]: I recommend to have decimal part in input values. Probably u will find smth. Apply to all

File: src/test/java/com/epam/cdp/testng/SqrtTest.java [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+    @Test(dataProvider = "valuesForSqrt")
+    public void firstSqrt(double first, double expected) {
+        double result = calc.sqrt(first);
+        Assert.assertEquals(result, expected, "incorrect result Sqrt");
+    }
+
+    @DataProvider(name = "valuesForSqrt")
+    public Object[][] dataForSum() {
+        return new Object[][]{
+                {4, 2.0},
```
Comment (tumashchick) [inline src/test/java/com/epam/cdp/testng/SqrtTest.java]: double check input values. There is a mistake

File: src/test/java/com/epam/cdp/junit/DivDoubleTest.java [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+        this.first = first;
+        this.second = second;
+        this.expected = expected;
+    }
+
+    @Test
+    public void FirstDivSecondDouble() {
+        double result = calc.div(first, second);
+        assertEquals(expected, result, 0.00005);
```
[the remark this answers is not in the document]
Comment (yauhen07) [reply src/test/java/com/epam/cdp/junit/DivDoubleTest.java]: done

File: src/test/java/com/epam/cdp/junit/ExeptionsTest.java (hunk 11 above)
Comment (yauhen07) [reply src/test/java/com/epam/cdp/junit/ExeptionsTest.java]: done

File: src/test/java/com/epam/cdp/junit/DivDoubleTest.java (hunk 12 above)
Comment (yauhen07) [reply src/test/java/com/epam/cdp/junit/DivDoubleTest.java]: done

File: src/test/resources/testng.xml [hunk 16]
```diff
@@ -0,0 +1,23 @@
+<!DOCTYPE suite SYSTEM "http://testng.org/testng-1.0.dtd">
+<suite name="Demo Suite" parallel="tests" thread-count="7">
+    <groups>
+        <run>
+            <exclude name="Traning"/>
```
[the remark this answers is not in the document]
Comment (yauhen07) [reply src/test/resources/testng.xml]: done

File: src/test/java/com/epam/cdp/junit/DivDoubleTest.java [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+    }
+
+    public DivDoubleTest(double first, double second, double expected) {
+        this.first = first;
+        this.second = second;
+        this.expected = expected;
+    }
+
+    @Test
+    public void FirstDivSecondDouble() {
```
[the remark this answers is not in the document]
Comment (yauhen07) [reply src/test/java/com/epam/cdp/junit/DivDoubleTest.java]: done

File: src/test/java/com/epam/cdp/junit/DivDoubleTest.java [hunk 18]
```diff
@@ -0,0 +1,42 @@
...
+        double result = calc.div(first, second);
+        assertEquals(expected, result, 0.00005);
+
```
[the remark this answers is not in the document]
Comment (yauhen07) [reply src/test/java/com/epam/cdp/junit/DivDoubleTest.java]: done

File: src/test/java/com/epam/cdp/junit/DivDoubleTest.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (yauhen07) [reply src/test/java/com/epam/cdp/junit/DivDoubleTest.java]: leave as is
